### PR TITLE
dispatcher: fix data race in the DefaultDispatchConcurrency

### LINKED ISSUE #42746: DATA RACE: TestSimple
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
current master
https://do.pingcap.net/jenkins/blue/organizations/jenkins/pingcap%2Ftidb%2Fghpr_unit_test/detail/ghpr_unit_test/21085/pipeline/

### 3. What did you see instead (Required)

```
[2023/04/03 11:28:25.817 +08:00] [INFO] [dispatcher.go:213] ["check task, subtasks aren't finished"] ["task ID"=1] [cnt=20]
    dispatcher_test.go:182: 
        	Error Trace:	disttask/framework/dispatcher/dispatcher_test.go:182
        	            				disttask/framework/dispatcher/dispatcher_test.go:239
        	Error:      	Not equal: 
        	            	expected: 20
        	            	actual  : 10
        	Test:       	TestSimpleNormalFlow
[2023/04/03 11:28:25.840 +08:00] [INFO] [dispatcher.go:236] ["detect task loop exits"] [error="context canceled"]
[2023/04/03 11:28:25.843 +08:00] [INFO] [dispatcher.go:148] ["dispatch task loop exits"] [error="context canceled"] [interval=30]
[2023/04/03 11:28:25.905 +08:00] [INFO] [dispatcher.go:213] ["check task, subtasks aren't finished"] ["task ID"=1] [cnt=20]
[2023/04/03 11:28:25.907 +08:00] [INFO] [dispatcher.go:252] ["detect task exits"] ["task ID"=1] [error="context canceled"]
==================
WARNING: DATA RACE
Read at 0x00000a1c7ed8 by goroutine 105320:
  github.com/pingcap/tidb/disttask/framework/dispatcher.(*dispatcher).DispatchTaskLoop()
      disttask/framework/dispatcher/dispatcher.go:152 +0x1f7
  github.com/pingcap/tidb/disttask/framework/dispatcher.(*dispatcher).DispatchTaskLoop-fm()
      <autogenerated>:1 +0x39
  github.com/pingcap/tidb/util.(*WaitGroupWrapper).Run.func1()
      util/wait_group_wrapper.go:154 +0x87

Previous write at 0x00000a1c7ed8 by goroutine 52136:
  github.com/pingcap/tidb/disttask/framework/dispatcher_test.checkDispatch.func1()
      disttask/framework/dispatcher/dispatcher_test.go:147 +0x71
  runtime.deferCallSave()
      GOROOT/src/runtime/panic.go:796 +0x87
  testing.(*T).FailNow()
      <autogenerated>:1 +0x37
  github.com/stretchr/testify/require.Equal()
      external/com_github_stretchr_testify/require/require.go:165 +0xeb
  github.com/pingcap/tidb/disttask/framework/dispatcher_test.checkDispatch()
      disttask/framework/dispatcher/dispatcher_test.go:182 +0x5e6
  github.com/pingcap/tidb/session.BootstrapSession()
      session/session.go:3324 +0x7cc
  github.com/pingcap/tidb/domain.(*Domain).GetSessionCache()
      domain/sysvar_cache.go:62 +0x71
  github.com/pingcap/tidb/session.(*session).loadCommonGlobalVariablesIfNeeded()
      session/session.go:3666 +0x11e
  github.com/pingcap/tidb/session.(*session).ExecuteStmt()
      session/session.go:2103 +0x16b
  github.com/pingcap/tidb/session.(*session).ExecuteInternal()
      session/session.go:1638 +0x365
  github.com/pingcap/tidb/domain.(*Domain).LoadPrivilegeLoop()
      domain/domain.go:1530 +0x143
  github.com/pingcap/tidb/session.BootstrapSession()
      session/session.go:3317 +0x76c
  github.com/pingcap/tidb/testkit.bootstrap()
      testkit/mockstore.go:85 +0x84
  github.com/pingcap/tidb/testkit.CreateMockStoreAndDomain()
      testkit/mockstore.go:70 +0xd0
  github.com/pingcap/tidb/testkit.CreateMockStore()
      testkit/mockstore.go:62 +0x3b2
  github.com/pingcap/tidb/disttask/framework/dispatcher_test.MockDispatcher()
      disttask/framework/dispatcher/dispatcher_test.go:38 +0x55
  github.com/pingcap/tidb/disttask/framework/dispatcher_test.checkDispatch()
      disttask/framework/dispatcher/dispatcher_test.go:141 +0xa6
  github.com/pingcap/tidb/disttask/framework/dispatcher_test.TestSimpleNormalFlow()
      disttask/framework/dispatcher/dispatcher_test.go:239 +0x38
  testing.tRunner()
      GOROOT/src/testing/testing.go:1576 +0x216
  testing.(*T).Run.func1()
      GOROOT/src/testing/testing.go:1629 +0x47

Goroutine 105320 (running) created at:
  github.com/pingcap/tidb/util.(*WaitGroupWrapper).Run()
      util/wait_group_wrapper.go:152 +0xf6
  github.com/pingcap/tidb/disttask/framework/dispatcher.(*dispatcher).Start()
      disttask/framework/dispatcher/dispatcher.go:129 +0xae
  github.com/pingcap/tidb/domain.(*Domain).distTaskFrameworkLoop.func2()
      domain/domain.go:1404 +0x2d0
  github.com/pingcap/tidb/domain.(*Domain).distTaskFrameworkLoop()
      domain/domain.go:1424 +0x321
  github.com/pingcap/tidb/domain.(*Domain).initDistTaskLoop.func1()
      domain/domain.go:1378 +0x177
  github.com/pingcap/tidb/util.(*WaitGroupEnhancedWrapper).Run.func1()
      util/wait_group_wrapper.go:96 +0xde

Goroutine 52136 (running) created at:
  testing.(*T).Run()
      GOROOT/src/testing/testing.go:1629 +0x805
  testing.runTests.func1()
      GOROOT/src/testing/testing.go:2036 +0x8d
  testing.tRunner()
      GOROOT/src/testing/testing.go:1576 +0x216
  testing.runTests()
      GOROOT/src/testing/testing.go:2034 +0x87c
  testing.(*M).Run()
      GOROOT/src/testing/testing.go:1906 +0xb44
  go.uber.org/goleak.VerifyTestMain()
      external/org_uber_go_goleak/testmain.go:53 +0x70
  github.com/pingcap/tidb/disttask/framework/dispatcher.TestMain()
      disttask/framework/dispatcher/main_test.go:49 +0x632
  main.main()
      bazel-out/k8-fastbuild/bin/disttask/framework/dispatcher/dispatcher_test_/testmain.go:126 +0x7ce
==================
```

### PR DESCRIPTION
**Pull request by hawkingrei**
### What problem does this PR solve?

Issue Number: close #42746

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by xhebox — APPROVED** [review]

**Review by lcwangchao — APPROVED** [review]

File: disttask/framework/dispatcher/dispatcher_test.go [hunk 1]
```diff
@@ -144,7 +143,7 @@ func checkDispatch(t *testing.T, taskCnt int, isSucc bool) {
 		dsp.Stop()
 		// make data race happy
 		if taskCnt == 1 {
-			dispatcher.DefaultDispatchConcurrency = originalConcurrency
```
Comment (zimulala) [inline disttask/framework/dispatcher/dispatcher_test.go:147]: I call this line after `dsp.Stop()`, it seems this won't encounter data race.  Or is it something else?

**Review by zimulala — CHANGES_REQUESTED** [review]

**Comment by hawkingrei** [comment]
It is not a good way to fix it.

**Pull request closed without merging**